File: tools/midi-tone/screensaver.py
#!/usr/bin/env python3
"""Idle blanking for the TFT kiosk — image-retention / burn-in protection.

The panel is an IPS LCD (BigTreeTech Pi TFT70). The kiosk chrome is bold and
mostly static, so a long jam with no touch still ghosts the same pixels.
midi-tone already turns X DPMS off so the kiosk never blanks on its own; this
module is the replacement:

- After a stretch of **no touch**, cover the UI and dim the DSI backlight.
- While the UI is up, slowly shift it by a couple of pixels so labels/boxes
  do not sit on the same cells for the whole session.

Audio / sequencer / songs keep running. Only a tap wakes the panel — MIDI
never counts as activity.
"""

from __future__ import annotations

import math
import os
import pathlib
from typing import Mapping, Optional, Sequence, Tuple


DEFAULT_TIMEOUT_SEC = 180.0  # 3 minutes
TIMEOUT_PRESETS: Tuple[float, ...] = (60.0, 180.0, 600.0, 0.0)
PIXEL_SHIFT_AMPLITUDE = 2
PIXEL_SHIFT_DWELL_SEC = 40.0


def timeout_from_env(env: Optional[Mapping[str, str]] = None) -> float:
    src = os.environ if env is None else env
    raw = src.get("MIDI_TONE_SCREENSAVER_SEC")
    if raw is None or raw.strip() == "":
        return DEFAULT_TIMEOUT_SEC
    try:
        return max(0.0, float(raw))
    except ValueError:
        return DEFAULT_TIMEOUT_SEC


def next_timeout_preset(current: float, presets: Sequence[float] = TIMEOUT_PRESETS) -> float:
    values = [float(p) for p in presets]
    cur = float(current)
    for i, value in enumerate(values):
        if abs(value - cur) < 0.5:
            return values[(i + 1) % len(values)]
    return values[0] if values else DEFAULT_TIMEOUT_SEC


def timeout_label(seconds: float) -> str:
    sec = float(seconds)
    if sec <= 0:
        return "BLANK OFF"
    minutes = sec / 60.0
    if abs(minutes - round(minutes)) < 0.05 and round(minutes) >= 1:
        return f"BLANK {int(round(minutes))} MIN"
    return f"BLANK {int(round(sec))}s"


def orbit_xy(
    elapsed: float,
    width: int,
    height: int,
    label_w: int,
    label_h: int,
    margin: int = 16,
) -> Tuple[int, int]:
    """Slow incommensurate Lissajous path so the wake hint never sits still."""
    span_x = max(0, int(width) - int(label_w) - 2 * margin)
    span_y = max(0, int(height) - int(label_h) - 2 * margin)
    # Periods 47s / 31s don't close quickly — classic pixel-shift screensaver.
    x = margin + int((math.sin(elapsed * 2.0 * math.pi / 47.0) * 0.5 + 0.5) * span_x)
    y = margin + int(
        (math.sin(elapsed * 2.0 * math.pi / 31.0 + 1.2) * 0.5 + 0.5) * span_y
    )
    return x, y


def pixel_shift_xy(
    elapsed: float,
    amplitude: int = PIXEL_SHIFT_AMPLITUDE,
    dwell_sec: float = PIXEL_SHIFT_DWELL_SEC,
) -> Tuple[int, int]:
    """1–2px square orbit. Dwells so the chrome is not constantly jittering."""
    amp = max(1, int(amplitude))
    dwell = dwell_sec if dwell_sec > 0 else PIXEL_SHIFT_DWELL_SEC
    idx = int(max(0.0, float(elapsed)) // dwell) % 8
    pattern = (
        (0, 0),
        (1, 0),
        (1, 1),
        (0, 1),
        (-1, 1),
        (-1, 0),
        (-1, -1),
        (0, -1),
    )
    dx, dy = pattern[idx]
    return dx * amp, dy * amp


class IdleWatch:
    """Monotonic idle timer. Tk must not be called from the MIDI thread."""

    def __init__(self, timeout_sec: float = DEFAULT_TIMEOUT_SEC, *, now: Optional[float] = None) -> None:
        self.timeout_sec = max(0.0, float(timeout_sec))
        self._last = 0.0 if now is None else float(now)
        if now is None:
            import time

            self._last = time.monotonic()
        self.active = False

    def poke(self, now: Optional[float] = None) -> bool:
        """Record activity. Returns True if this dismissed an active saver."""
        import time

        self._last = time.monotonic() if now is None else float(now)
        if self.active:
            self.active = False
            return True
        return False

    def due(self, now: Optional[float] = None) -> bool:
        if self.active or self.timeout_sec <= 0:
            return False
        import time

        t = time.monotonic() if now is None else float(now)
        return (t - self._last) >= self.timeout_sec

    def activate(self) -> bool:
        """Mark saver on. Returns False if it was already showing."""
        if self.active:
            return False
        self.active = True
        return True


class PanelBacklight:
    """Best-effort DSI backlight via /sys/class/backlight (needs video group)."""

    def __init__(self, root: Optional[pathlib.Path] = None) -> None:
        self.root = pathlib.Path("/sys/class/backlight") if root is None else pathlib.Path(root)
        self._path: Optional[pathlib.Path] = None
        self._saved: Optional[int] = None
        self._power_path: Optional[pathlib.Path] = None
        self._saved_power: Optional[int] = None

    def _find(self) -> Optional[pathlib.Path]:
        if self._path is not None and self._path.is_file():
            return self._path
        if not self.root.is_dir():
            return None
        candidates = sorted(p for p in self.root.glob("*/brightness") if p.is_file())
        if not candidates:
            return None
        self._path = candidates[0]
        power = self._path.parent / "bl_power"
        self._power_path = power if power.is_file() else None
        return self._path

    def dim(self) -> bool:
        path = self._find()
        if path is None:
            return False
        try:
            current = int(path.read_text(encoding="ascii").strip() or "0")
        except (OSError, ValueError):
            return False
        if self._saved is None:
            self._saved = current
        if self._power_path is not None and self._saved_power is None:
            try:
                self._saved_power = int(
                    self._power_path.read_text(encoding="ascii").strip() or "0"
                )
            except (OSError, ValueError):
                self._saved_power = None
        ok = False
        try:
            path.write_text("0\n", encoding="ascii")
            ok = True
        except OSError:
            pass
        if self._power_path is not None:
            try:
                # 4 = FB_BLANK_POWERDOWN
                self._power_path.write_text("4\n", encoding="ascii")
                ok = True
            except OSError:
                pass
        return ok

    def restore(self) -> bool:
        path = self._path
        if path is None:
            return False
        ok = False
        if self._power_path is not None:
            value = 0 if self._saved_power is None else int(self._saved_power)
            try:
                self._power_path.write_text(f"{value}\n", encoding="ascii")
                ok = True
            except OSError:
                pass
        if self._saved is not None:
            try:
                path.write_text(f"{int(self._saved)}\n", encoding="ascii")
                ok = True
            except OSError:
                pass
        self._saved = None
        self._saved_power = None
        return ok

File: crates/pidi-native/src/screensaver.rs
//! TFT idle blanking and burn-in protection (parity with `apps/pidi/pidi/screensaver.py`).

use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

pub const DEFAULT_TIMEOUT_SEC: f32 = 180.0;
pub const TIMEOUT_PRESETS: [f32; 4] = [60.0, 180.0, 600.0, 0.0];
pub const PIXEL_SHIFT_AMPLITUDE: i32 = 2;
pub const PIXEL_SHIFT_DWELL_SEC: f32 = 40.0;

pub fn timeout_from_env() -> f32 {
    std::env::var("MIDI_TONE_SCREENSAVER_SEC")
        .ok()
        .and_then(|raw| raw.trim().parse::<f32>().ok())
        .map(|v| v.max(0.0))
        .unwrap_or(DEFAULT_TIMEOUT_SEC)
}

pub fn next_timeout_preset(current: f32) -> f32 {
    for (i, value) in TIMEOUT_PRESETS.iter().enumerate() {
        if (*value - current).abs() < 0.5 {
            return TIMEOUT_PRESETS[(i + 1) % TIMEOUT_PRESETS.len()];
        }
    }
    TIMEOUT_PRESETS[0]
}

pub fn timeout_label(seconds: f32) -> &'static str {
    if seconds <= 0.0 {
        return "BLANK OFF";
    }
    let minutes = seconds / 60.0;
    if (minutes - minutes.round()).abs() < 0.05 && minutes.round() >= 1.0 {
        return match minutes.round() as i32 {
            1 => "BLANK 1 MIN",
            3 => "BLANK 3 MIN",
            10 => "BLANK 10 MIN",
            n => {
                let _ = n;
                "BLANK MIN"
            }
        };
    }
    match seconds.round() as i32 {
        60 => "BLANK 1 MIN",
        180 => "BLANK 3 MIN",
        600 => "BLANK 10 MIN",
        n if n > 0 => "BLANK SEC",
        _ => "BLANK OFF",
    }
}

pub fn timeout_label_dynamic(seconds: f32) -> String {
    if seconds <= 0.0 {
        return "BLANK OFF".into();
    }
    let minutes = seconds / 60.0;
    if (minutes - minutes.round()).abs() < 0.05 && minutes.round() >= 1.0 {
        return format!("BLANK {} MIN", minutes.round() as i32);
    }
    format!("BLANK {}s", seconds.round() as i32)
}

pub fn pixel_shift_xy(elapsed: f32) -> (i32, i32) {
    let amp = PIXEL_SHIFT_AMPLITUDE.max(1);
    let dwell = PIXEL_SHIFT_DWELL_SEC.max(0.1);
    let idx = (elapsed.max(0.0) / dwell) as usize % 8;
    let pattern = [
        (0, 0),
        (1, 0),
        (1, 1),
        (0, 1),
        (-1, 1),
        (-1, 0),
        (-1, -1),
        (0, -1),
    ];
    let (dx, dy) = pattern[idx];
    (dx * amp, dy * amp)
}

pub struct IdleWatch {
    pub timeout_sec: f32,
    last_activity: Instant,
    pub active: bool,
}

impl IdleWatch {
    pub fn new(timeout_sec: f32) -> Self {
        Self {
            timeout_sec: timeout_sec.max(0.0),
            last_activity: Instant::now(),
            active: false,
        }
    }

    pub fn poke(&mut self) -> bool {
        self.last_activity = Instant::now();
        if self.active {
            self.active = false;
            return true;
        }
        false
    }

    pub fn due(&self) -> bool {
        if self.active || self.timeout_sec <= 0.0 {
            return false;
        }
        self.last_activity.elapsed() >= Duration::from_secs_f32(self.timeout_sec)
    }

    pub fn activate(&mut self) -> bool {
        if self.active {
            return false;
        }
        self.active = true;
        true
    }
}

pub struct PanelBacklight {
    root: PathBuf,
    brightness: Option<PathBuf>,
    saved_brightness: Option<u32>,
}

impl Default for PanelBacklight {
    fn default() -> Self {
        Self::new()
    }
}

impl PanelBacklight {
    pub fn new() -> Self {
        Self {
            root: PathBuf::from("/sys/class/backlight"),
            brightness: None,
            saved_brightness: None,
        }
    }

    fn find_brightness(&mut self) -> Option<&Path> {
        if self
            .brightness
            .as_ref()
            .is_some_and(|p| p.is_file())
        {
            return self.brightness.as_deref();
        }
        let Ok(entries) = std::fs::read_dir(&self.root) else {
            return None;
        };
        let mut candidates: Vec<PathBuf> = entries
            .flatten()
            .map(|e| e.path().join("brightness"))
            .filter(|p| p.is_file())
            .collect();
        candidates.sort();
        self.brightness = candidates.into_iter().next();
        self.brightness.as_deref()
    }

    pub fn dim(&mut self) -> bool {
        let Some(path) = self.find_brightness().map(Path::to_path_buf) else {
            return false;
        };
        let current = std::fs::read_to_string(&path)
            .ok()
            .and_then(|s| s.trim().parse::<u32>().ok())
            .unwrap_or(0);
        if self.saved_brightness.is_none() {
            self.saved_brightness = Some(if current > 0 {
                current
            } else {
                self.max_brightness(&path)
            });
        }
        std::fs::write(&path, "0\n").is_ok()
    }

    pub fn restore(&mut self) -> bool {
        let Some(path) = self.find_brightness().map(Path::to_path_buf) else {
            return false;
        };
        let power = path.parent().map(|p| p.join("bl_power"));
        if let Some(power) = power.as_ref().filter(|p| p.is_file()) {
            let _ = std::fs::write(power, "0\n");
        }
        let value = self
            .saved_brightness
            .filter(|v| *v > 0)
            .unwrap_or_else(|| self.max_brightness(&path));
        let ok = std::fs::write(&path, format!("{value}\n")).is_ok();
        self.saved_brightness = None;
        ok
    }

    fn max_brightness(&self, path: &Path) -> u32 {
        path.parent()
            .and_then(|p| {
                std::fs::read_to_string(p.join("max_brightness"))
                    .ok()
                    .and_then(|s| s.trim().parse::<u32>().ok())
            })
            .filter(|v| *v > 0)
            .unwrap_or(255)
    }

    pub fn ensure_lit(&mut self) -> bool {
        let Some(path) = self.find_brightness().map(Path::to_path_buf) else {
            return false;
        };
        let current = std::fs::read_to_string(&path)
            .ok()
            .and_then(|s| s.trim().parse::<u32>().ok())
            .unwrap_or(0);
        if current > 0 {
            return true;
        }
        self.restore()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn timeout_presets_cycle() {
        assert_eq!(next_timeout_preset(180.0), 600.0);
        assert_eq!(next_timeout_preset(600.0), 0.0);
        assert_eq!(next_timeout_preset(0.0), 60.0);
    }

    #[test]
    fn pixel_shift_steps() {
        assert_eq!(pixel_shift_xy(0.0), (0, 0));
        assert_eq!(pixel_shift_xy(39.9), (0, 0));
        assert_eq!(pixel_shift_xy(40.0), (2, 0));
    }
}
